pkg/kv/util/weak_owner.go

```diff
@@ -0,0 +1,275 @@
+					"new_owner": ownership.Owner,
+				}).Info("Lost ownership race")
+				break
+			}
+			expires := time.Now().Add(w.refreshInterval)
+			ownership.Expires = timestamppb.New(expires)
+			err = kv.SetMsgIf(ctx, w.Store, weakOwnerPartition, prefixedKey, &ownership, predicate)
+			if err != nil {
+				log.WithError(err).Warn("Failed to set ownership message to refresh (keep going, may lose)")
+				continue
+			}
+			log.WithField("expires", expires).Trace("Refreshed ownership")
+		}
+	}
+}
+
+// sleepFor sleeps for duration or until ctx is Done.  If Done, it returns
+// the error from ctx or finished.
+//
+// TODO(ariels): Rewrite once on Go 1.23, which no longer leaks time.After.
+func sleepFor(ctx context.Context, duration time.Duration) error {
+	timer := time.NewTimer(duration)
+	defer timer.Stop()
+	for {
+		select {
+		case <-timer.C:
+			return nil
+		case <-ctx.Done():
+			err := ctx.Err()
+			if err == nil {
+				err = finished
+			}
+			return err
+		}
+	}
+}
+
+type keyOwnership string
+
+const (
+	keyOwned    keyOwnership = "owned"
+	keyNotOwned keyOwnership = "not owned"
+	keyExpired  keyOwnership = "expired"
+)
+
+func checkOwnership(expires, now time.Time, getErr error) keyOwnership {
+	switch {
+	case errors.Is(getErr, kv.ErrNotFound):
+		return keyNotOwned
+	case expires.Before(now):
+		return keyExpired
+	default:
+		return keyOwned
+	}
+}
+
+// startOwningKey blocks until it gets weak ownership of key in store.
+// This is a spin-wait (with sleeps) because of the KV interface.
+//
+// TODO(ariels): Spin only once (maybe use WaitFor, or chain requests and
+// spin only on the first) when multiple goroutines all wait for the same
+// key.
+//
+// TODO(ariels): Be fair, at least in the same process.  Chaining requests
+// and spinning only on the first would do this as well.
+func (w *WeakOwner) startOwningKey(ctx context.Context, owner string, prefixedKey []byte) error {
+	log := w.Log.WithContext(ctx).WithFields(logging.Fields{
+		"owner":            owner,
+		"prefixed_key":     string(prefixedKey),
+		"acquire_interval": w.acquireInterval,
+		"refresh_interval": w.refreshInterval,
+	})
+	for {
+		ownership := WeakOwnership{}
+		predicate, err := kv.GetMsg(ctx, w.Store, weakOwnerPartition, prefixedKey, &ownership)
+		if err != nil && !errors.Is(err, kv.ErrNotFound) {
+			return fmt.Errorf("start owning %s for %s: %w", prefixedKey, owner, err)
+		}
+
+		now := time.Now()
+		free := checkOwnership(ownership.Expires.AsTime(), now, err)
+		if free != keyOwned {
+			expiryTime := now.Add(w.refreshInterval)
+			ownership = WeakOwnership{
+				Owner:   owner,
+				Expires: timestamppb.New(expiryTime),
+				Comment: fmt.Sprintf("%s@%v", owner, now),
+			}
+			log.WithFields(logging.Fields{
+				"ownership": free,
+				"expires":   expiryTime,
+				"now":       now,
+			}).Trace("Try to take ownership")
+			if errors.Is(err, kv.ErrNotFound) {
+				err = kv.SetMsg(ctx, w.Store, weakOwnerPartition, prefixedKey, &ownership)
```

Comment:
This is wrong, no? multiple routines can succeed at the same time

Reply:
Ouch, yeah, this was supposed to be a SetIf with the "not present" predicate - and I forgot to find out the correct value for that predicate.
Thanks for catching this!